common/src/main/java/org/apache/nemo/common/ir/vertex/utility/TaskSizeSplitterVertex.java

```diff
@@ -0,0 +1,345 @@
+        final IREdge newIrEdge = new IREdge(edge.getPropertyValue(CommunicationPatternProperty.class).get(),
+          originalToNewIRVertex.get(edge.getSrc()), dstVertex);
+        edge.copyExecutionPropertiesTo(newIrEdge);
+        this.addDagIncomingEdge(newIrEdge);
+      }));
+    }
+
+    increaseTestingTrial();
+    return this;
+  }
+
+  // private helper methods
+
+  /**
+   * Set Parallelism Property of internal vertices by unroll iteration.
+   * @param irVertex    vertex to set parallelism property.
+   */
+  private void setParallelismPropertyByTestingTrial(final IRVertex irVertex) {
+    if (testingTrial == 0 && !(irVertex instanceof OperatorVertex
+      && ((OperatorVertex) irVertex).getTransform() instanceof SignalTransform)) {
+      irVertex.setPropertyPermanently(ParallelismProperty.of(32));
+    } else {
+      irVertex.setProperty(ParallelismProperty.of(1));
+    }
+  }
+
+  /**
+   * Set SubPartitionSetProperty of given edge by unroll iteration.
+   * @param edge    edge to set subPartitionSetProperty
+   */
+  private void setSubPartitionSetPropertyByTestingTrial(final IREdge edge) {
+    final ArrayList<KeyRange> partitionSet = new ArrayList<>();
+    int taskIndex = 0;
+    if (testingTrial == 0) {
+      for (int i = 0; i < 4; i++) {
+        partitionSet.add(taskIndex, HashRange.of(i, i + 1));
+        taskIndex++;
+      }
+      for (int groupStartingIndex = 4; groupStartingIndex < 512; groupStartingIndex *= 2) {
+        int growingFactor = groupStartingIndex / 4;
+        for (int startIndex = groupStartingIndex; startIndex < groupStartingIndex * 2; startIndex += growingFactor) {
+          partitionSet.add(taskIndex, HashRange.of(startIndex, startIndex + growingFactor));
+          taskIndex++;
+        }
+      }
+      edge.setProperty(SubPartitionSetProperty.of(partitionSet));
+    } else {
+      partitionSet.add(0, HashRange.of(512, partitionerProperty)); // 31+testingTrial
+      edge.setProperty(SubPartitionSetProperty.of(partitionSet));
+    }
+  }
+
+  /**
+   * Mark edges for DTS (i.e. incoming edges of second iteration vertices).
+   * @param toAssign          Signal Vertex to get MessageIdVertexProperty
+   * @param edgesToOptimize   Edges to mark for DTS
+   */
+  private void markEdgesToOptimize(final List<OperatorVertex> toAssign, final Set<IREdge> edgesToOptimize) {
+    if (testingTrial > 0) {
+      edgesToOptimize.forEach(edge -> {
+        if (!edge.getDst().getPropertyValue(ParallelismProperty.class).get().equals(1)) {
+          throw new IllegalArgumentException("Target edges should begin with Parallelism of 1.");
+        }
+        final HashSet<Integer> msgEdgeIds =
+          edge.getPropertyValue(MessageIdEdgeProperty.class).orElse(new HashSet<>(0));
+        msgEdgeIds.add(toAssign.get(0).getPropertyValue(MessageIdVertexProperty.class).get());
+        edge.setProperty(MessageIdEdgeProperty.of(msgEdgeIds));
+      });
+    }
+  }
+
+  public void printLogs() {
```

Comment:
override toString()?
https://www.tutorialspoint.com/java/number_tostring.htm

please refer to other classes in the repository